### PR TITLE
Add optional dependency support.

### LINKED ISSUE #7: Support plugin dependencies
I've been using [vim-addon-manager](https://github.com/MarcWeber/vim-addon-manager), which does more or less what this addon seems to do.  I've not used vundle yet, so I'm not sure how similar the two addons are, but VAM does provide the same core functionality of allowing addons to be installed and enabled from within vim itself.  VAM is a bit sprawling, partly because it also does the work of vim-scripts.org itself, allowing addons to be pulled in directly from vim.org or from git, SVN, hg, etc. repos.  So it's nice to see this more concise addon being made available as a potential replacement.

One thing that VAM provides that I don't see in vundle is the ability to declare addon dependencies.  Each addon can have a file called `${ADDON_NAME}-addon-info.txt` in the root of its directory structure.  This text file is in vim-compatible JSON format and allows various metadata, including dependencies on other addons, to be declared.  To me, this is a must-have feature, since it provides a way to modularize complex VimL code into libraries, allowing more complex functionality to be shared among addons and reducing code duplication and memory footprints.

Is there a possibility that this feature might be worked into `vundle`?  If so, could a compatible metadata format be used?

### LINKED ISSUE #384: Remove dependency management from the todo
I'm suggesting that we remove "dependencies" from the Vundle TODO list in the README file and decide that dependency management is something that should be handled by more advanced plugin managers like VAM and NeoBundle.

I don't like the idea that if I tell Vundle to install one plugin it installs 2 more dependencies. If one plugin depends on another it will have to tell me and I'll make a choice to include it into my vimrc. The vimrc will actually mirror the installed plugins and I'll know for sure which plugins are installed.

### PR DESCRIPTION
**Pull request by bcnjr5**
- Closes #7
 - Closes #384

### CODE REVIEW

**Comment by jdevera** [comment]
This PR dance is becoming a bit confusing. I'd be happy to test a change that will introduce some very crude compatibility with the addon-info.json files as long as I can find some formal description of the format (I'll look for one) and @gmarik gives us the go ahead in this direction.

The other file format, as I stated before, seems to me a new invention and I don't wish to promote it by including it in Vundle.

So I'll cherry pick the relevant commit and give it a spin for a few weeks.

File: autoload/vundle/scripts.vim [hunk 1]
```diff
@@ -254,4 +254,36 @@ func! s:load_scripts(bang)
...
+" Get the dependencies of the given bundle.
+"
+" bundle -- The bundle to check. (must be installed)
+" return -- List of bundle names.
+" ---------------------------------------------------------------------------
+func! vundle#scripts#getdeps(bundle)
+  if filereadable(a:bundle['rtpath'] . '/depends.txt')
+    return eval("['" . join(readfile(a:bundle['rtpath'] . '/depends.txt'), "','") . "']")
+  elseif filereadable(a:bundle['rtpath'] . '/addon-info.json')
+    let true = 1
```
Comment (jdevera) [inline autoload/vundle/scripts.vim]: These true, false and null are not used, why are they here?

File: autoload/vundle/scripts.vim [hunk 2]
```diff
@@ -254,4 +254,36 @@ func! s:load_scripts(bang)
...
+
+" ---------------------------------------------------------------------------
+" Get the dependencies of the given bundle.
+"
+" bundle -- The bundle to check. (must be installed)
+" return -- List of bundle names.
+" ---------------------------------------------------------------------------
+func! vundle#scripts#getdeps(bundle)
+  if filereadable(a:bundle['rtpath'] . '/depends.txt')
+    return eval("['" . join(readfile(a:bundle['rtpath'] . '/depends.txt'), "','") . "']")
```
Comment (jdevera) [inline autoload/vundle/scripts.vim]: You know, I suggested the removal of this particular item, but in any case, let me note that you start with an array of lines, then join it with commas and quotes, and then eval it, so you end up with... an array of lines.

**Comment by bcnjr5** [comment]
@jdevera:
 - The `depends.txt` is better than nothing, you can remove it if you want.
 - The `true`, `false`, and `null` variables are constants in JSON, so when vim evalulates JSON code as vim script, it needs to know what to do if it encounters one.
 - Calling my git URL support "crude" is an understatement. I wish it was, but I think the term "hack" (or maybe "kludge") is more suitable. :wink:

File: autoload/vundle/scripts.vim (hunk 1 above)
Comment (dbarnett) [inline autoload/vundle/scripts.vim]: Other people will probably have the same question. You should mention in a comment that they're defined for use inside the eval.

**Comment by bcnjr5** [comment]
Here you go.

**Comment by jdevera** [comment]
Thanks for the changes.

Now, I was thinking about this change yesterday and the security aware part of my brain suddenly woke up and cried *WAIT.A.SECOND.*: We parse JSON by eval'ing it. Vundle does this for the list of Vundles, which comes from a hardcoded URL, which is less bad, but still bad. Now we are talking about eval'ing code that we get from arbitrary sources, so any plugin author could have some *interesting* :wink: :wink:  code in their JSON file and we'd be running it happily.

I think we need to think this through a bit better, and maybe find a less risky way of getting those plugins from the JSON file.

**Comment by jdevera** [comment]
Some relevant info:
http://stackoverflow.com/questions/17751186/iterating-over-a-string-in-vimscript-or-parse-a-json-file

**Comment by bcnjr5** [comment]
@jdevera: I had wondered the same thing myself, but [that is the way it is done in vim-addon-manager](https://github.com/MarcWeber/vim-addon-manager/blob/master/autoload/vam.vim) (line 137)

Let's see if we can't make it safer...

**Comment by dbarnett** [comment]
@bcnjr5 You missed the `vam#VerifyIsJson` part where vim-addon-manager validates it with a regex first, though. Now I think about it, there must be a hack since a regex couldn't represent arbitrarily deep nesting of JSON data structures, but that's at least not a security hole (as long as you can trust the regex to only match safe JSON).

**Comment by bcnjr5** [comment]
@dbarnett: `vam#VerifyIsJson` seems to be meant for checking syntax, not embedded vim script. Do you think copying that function over would help? I think vim will fail to `eval` JSON with syntax errors.

**Comment by bcnjr5** [comment]
The issue of ensuring JSON is safe seems to go against the unix principle of blindly doing *exactly* what the user says. Furthermore, I have yet to find a method of safely parsing JSON that isn't full of warnings about how slow it is.

**Comment by dbarnett** [comment]
> Do you think copying that function over would help?

Yes. If you validate that a given string is valid JSON syntax (or some subset thereof), then you've also ensured that it's safe to eval since there is no string that is both valid JSON and malicious vimscript code. A real parser might be slow, but that has nothing to do with matching the string against a regex before evaling it.

That said, I wouldn't trust that bit of code to be 100% secure without first understanding exactly how it works and ensuring that it's impossible for it to match malicious vimscript code. But either way I certainly wouldn't eval strings without making any attempt to validate them first. I guarantee "blindly execute strings as code" does not fit the unix principles, in particular the "Rule of Repair" which says "fail in a manner that is easy to localize and diagnose".

If there is a real JSON parser available that would be best and I seriously seriously doubt parser performance is going to be significant compared to filesystem operations with the small simple JSON files we're talking about in practice.

**Comment by Shougo** [comment]
I think to use `:sandbox` is better.

**Comment by jdevera** [comment]
Hey @Shougo that looks promising!

**Comment by bcnjr5** [comment]
(Sorry for the delay.)

Yup. `:sandbox` will work nicely. Thanks, @Shougo!

**Comment by bcnjr5** [comment]
At this rate, I am 30% sure vim has all the functionality I added built in already.

**Comment by bcnjr5** [comment]
@dbarnett:
> fail in a manner that is easy to localize and diagnose

```
Segmentation fault (core dumped)
```

I wish. :confused:

**Comment by bcnjr5** [comment]
@jdevera: Could you provide an example of malicious vim script embedded in a JSON file?

**Comment by av3r4ge** [comment]
Not much of a json, but if you eval it, it works: `system('rm tst')`

**Comment by dbarnett** [comment]
I agree using `:sandbox` should prevent any actually nasty side effects.

Not sure what you want examples for, but something like this could get through `eval()` undetected and also evaluate to a reasonable value:
```vim
system('rm tst') ? {} : {
  "dependencies": {}
}
```

**Comment by bcnjr5** [comment]
Testing them on a VM to ensure nothing bad happens.

**Comment by bcnjr5** [comment]
Everything is safe (as far as I know) except for the calls to `system()`: my vim doesn't support it or something. The function is mentioned with `:help system`, but is undefined. I probably need to set some option or something first, so I need someone to either tell me what I'm doing wrong this time, or test the above cases on their machine.

If the `system()` function requires some option to be set, it may be a good idea to temporarily unset that option before the `sandbox eval()` call.

(My vim is `vim-nox` on Debian jessie, by the way. (Running in a huge kludge on a Nook Color. (armhf)))

**Comment by av3r4ge** [comment]
Weird. `system()` should be set, but it builds its command from several options. Maybe they aren't set or somthing.

Sandbox, prevents executing shell commands, among other things, so that should be fine.

But, this still need to be thought of. What would happen if, for example, the name of the package contains `../`?
Does that make it installed in some arbitrary place?
What happens if we accidentally clear such a thing? Does it `rm -rf` the entire hierarchy of directories?

**Comment by bcnjr5** [comment]
@av3r4ge: That is a ***very*** good question, but I think that should be taken up with Vundle itself, not just the dependency management portion. I'm not sure if that's even possible with git repository names or vim-scripts names. As we are using Vundle to process all dependency names instead of reinventing the wheel, I believe that no _new_ bugs/vulnerabilities are introduced.

**Comment by jdevera** [comment]
@bcnjr5 try running this:

```viml
:echo system('ls /')
```

**Comment by av3r4ge** [comment]
@bcnjr5 The problem starts when you are working with unknown data.
The users write their .vimrc files and there's no reason for them to enter weird plugin lines.
When you are working with unknown files... That's a differenct story.

For example, try to put `Plugin '../../vim-scripts/vundle'` in your .vimrc. You'll find a new directory in your home directory (if you work with the defaults on Linux). Not that bad, but it's not supposed to happen.
Another example is the following:
I use Linux, so I cloned a bare repository of vundle somewhere on my computer, changed it's name to `Vundle.vim\..\Vundle.vim` and updated it's server info to make it a nice server repo.
Then, I went to another computer with Windows and set up vim, git and vundle.
After everything was set, I added `Plugin 'http://my_comp/Vundle.vim\..\Vundle.vim` and `PluginInstall`ed. The result was a bunch of errors, but after reopening vim, everything was fine.
When I went to the Vundle.vim directory and typed `git remote -v` I saw my hijacked vundle version.

I'm not saying the dependency management is bad. I'm saying that when you insert unknown inputs to the equation, extra caution steps should be taken which previously might not have been as necessary.

**Comment by bcnjr5** [comment]
@av3r4ge: Yet another uncannily astute observation. Perhaps I should add another option?

`g:vundle_secure` (set to 1 by default)

Unless `g:vundle_secure` is set to `0`, dependencies will be rejected with an error if they include `./` or `../` in the name.

Then again, this form of dependency support is incomplete: it only supports literal git URLs. (for now)

PS: You would be very successful in one of those "bug bounty" programs.

**Comment by bcnjr5** [comment]
@dbarnett: Is it alright if the actual code is complex, yet easy to read and maintain, as long as the actual interface is KISS?

**Comment by av3r4ge** [comment]
@bcnjr5 First of all, Thanks :).
That's a good suggestion in the right direction. But I think a more robust solution would be whitelisting names instead of blacklisting them. I don't think a normal name should contain weird characters. Maybe a name should conform to something like `^([A-Za-z0-9_-]|\.[^\.])+$`?
Please tell if you know of any weird-named plugins which doesn't meet this.

Moreover, some fixes can be applied to Vundle itself. For instance, `s:expand_path` in autoload/vundle/config.vim expands the `g:bundle_dir` with the plugin's name. I think this is bad, because there's no reason for the name itself to contain expandable data (e.g. a plugin named $HOME).

Another suggestion that can be included for the more paranoids among us, is setting `g:vundle_no_deps = 1` by default and add an option like `{'deps': 1}` for specific plugins which will install the dependencies of just that single plugins (and it's dependencies, I guess).
That way, if I truly want a plugin with it's dependencies and decided I trust this chain, let it be.

What do you think?

**Comment by dbarnett** [comment]
Dependencies should never look like relative paths. Right now we're only talking about git URLs and anything that doesn't look like a git URL should be rejected (some validation could be added there depending on what Vundle currently does with it). It's reasonable to not want additional plugins installed without warning. What if Vundle by default prompted when a new dependency is encountered?

I could also see some users wanting to lock it down so only whitelisted plugins could install deps. From the comments on #7, there are at least a few vocal anti-dependency users in the Vundle community.

I'm not totally sure what you were asking about code complexity, but I'm not a Vundle maintainer so I don't think it's really a question for me to answer.

**Comment by bcnjr5** [comment]
@dbarnett: Issue #7 is already taken care of: `let g:vundle_no_deps = 1`

**Comment by bcnjr5** [comment]
@av3r4ge: How about rewriting it so you use `Plugin <name> -deps` to install dependencies of a plugin? That way, dependencies would be ignored by default, but easy to enable in a secure way.

I did consider adding a prompt for dependencies, but that would get annoying really fast, as it would either come up every time you launched vim, or make tasks that used to be automated require user input.

**Comment by av3r4ge** [comment]
@bcnjr5 Good idea. That's exactly what I suggeseted too (with a little notation difference) here:
> add an option like {'deps': 1} for specific plugins which will install the dependencies of just that single plugin (and it's dependencies, I guess).
> That way, if I truly want a plugin with it's dependencies and decided I trust this chain, let it be.

Vundle already has additional-options mechanism (rtp, name, pinned), so you might want to use that.

Before it hits main, we should consider hardening the naming convention and expanding of plugins' names (paths). This can come as part of this PR (as it requires Vundle to be a bit more secure in those areas), or as a seperate one.
@jdevera Would be nice to hear your opinion on the matter.

**Comment by dbarnett** [comment]
> I did consider adding a prompt for dependencies, but that would get annoying really fast, as it would either come up every time you launched vim, or make tasks that used to be automated require user input.

Not every time. When Vundle first encounters a dependency, before fetching it from git.

In vim-addon-manager, you install plugins interactively so you're more-or-less prompted for each one. For security, embedding malicious code in addon-info.json is made tricky if not impossible with a regex-based check. I don't think there's anything besides that.

**Comment by bcnjr5** [comment]
@dbarnett: How about [this](https://github.com/bcnjr5/Vundle.vim/commit/8d69f096878eea37c4ac23de6756c2041983f65b)?

**Comment by bcnjr5** [comment]
@jdevera: Which option do you think is best?

 - Parse dependencies of a plugin only if a certain argument is passed to the `Plugin <name>` line
 - Prompt the user when the dependency file is encountered for the first time

Additionally, some people might want do do some benign tricks like checking for a certain patch or version (`if has(...)`) and choosing which lines in the JSON file will be evaluated.

For example:

```vim
{
  "dependencies": {
    (has(gui)? "gui_plugin":"terminal_plugin"): {}
  }
}
```

Should we allow that?

**Comment by bcnjr5** [comment]
It also just occured to me that since Vundle uses a simple syntax for adding plugins instead of trying to copy VAM's complex, JSON-like plugin syntax; would it not be a good idea to also use a simple plugin-info syntax instead of trying to copy VAM's complex, JSON-like plugin-info syntax?

**Comment by dbarnett** [comment]
> How about [this](https://github.com/bcnjr5/Vundle.vim/commit/8d69f096878eea37c4ac23de6756c2041983f65b) for plugin names?

It doesn't make sense to me why checking `dep_name` would have anything to do with security when you're only using `dep_info.type` and `dep_info.url` to determine any behavior. I thought you just wanted to enforce `url` wasn't somehow a relative path that would confuse Vundle, and again I don't see any reason you would ever want to support a plugin declaring that it depends on a relative path on your system so I don't see the point of having an override.

I thought we were still talking about the security implications of parsing the addon-info.json. The other issue of transitively trusting dependencies is definitely not something regexes could help with. =/

> Additionally, some people might want do do some benign tricks …

Embedding vimscript in "JSON" is not a benign trick. The [vimdoc](https://github.com/google/vimdoc) tool deserializes the JSON files from python and expects them to be legitimate JSON. If someone ever needed that sort of logic, they should push to have it officially supported in the addon-info.json spec. If it's not feasible to prevent such ugliness in practice that's one thing, but it's not a feature.

**Comment by dbarnett** [comment]
> would it not be a good idea to also use a simple plugin-info syntax instead of trying to copy VAM's complex, JSON-like plugin-info syntax?

So back to the non-standard file listing Vundle names one per line? It would be easier to implement for Vundle, sure, but I can assure you I don't plan on duplicating my list of dependencies into two files with slightly different syntax for each of the plugins I maintain.

**Comment by bcnjr5** [comment]
@dbarnett:
> It doesn't make sense to me why checking dep_name would have anything to do with security when you're only using dep_info.type and dep_info.url to determine any behavior. I thought you just wanted to enforce url wasn't somehow a relative path that would confuse Vundle, and again I don't see any reason you would ever want to support a plugin declaring that it depends on a relative path on your system so I don't see the point of having an override.

I can't see any way of adding a relative path to a git URL. It will always be evaluated to just the repository name, right? The repository name is the name of a folder, not including the path _to_ that folder.

Checking the plugin name would simply be for posterity, or forward-compatibility. Would the method of checking the name be satisfactory, though? (It could be applied to other things than just the name.)

> Embedding vimscript in "JSON" is not a benign trick. The vimdoc tool deserializes the JSON files from python and expects them to be legitimate JSON. If someone ever needed that sort of logic, they should push to have it officially supported in the addon-info.json spec. If it's not feasible to prevent such ugliness in practice that's one thing, but it's not a feature.

Alright. Just a thought.

**Comment by bcnjr5** [comment]
@dbarnett:
> So back to the non-standard file listing Vundle names one per line?

Just putting everything on the table.

> I can assure you I don't plan on duplicating my list of dependencies into two files with slightly different syntax for each of the plugins I maintain.

That's the biggest problem with that idea.

**Comment by dbarnett** [comment]
> Checking the plugin name would simply be for posterity, or forward-compatibility.

Right, I just don't see why it would ever matter. For addon-info.json files, this name would only ever be a key into the vim-pi index and never passed to Vundle as a path.

**Comment by jdevera** [comment]
Apologies for my lack of answers. I'm currently in crunch time at work and getting zero time for github. Please give me a few days to digest these updates.

This PR is becoming a goldmine :)

**Comment by av3r4ge** [comment]
@bcnjr5 
> I can't see any way of adding a relative path to a git URL. It will always be evaluated to just the repository name, right? The repository name is the name of a folder, not including the path to that folder.

That's not entirely true.
Currently, Vundle choose the name of plugins as the last part of the uri splitted by '/', but for vim-scripts' plugins that's different. The name is the string passed to Plugin and the uri is build from that.
It's important to remember that windows' path separator is '\' and not '/'.
Furthermore, because the name is expanded, you could use something like `$HOME` as the name as well.

**Comment by av3r4ge** [comment]
@bcnjr5 @dbarnett @jdevera 
I've made a small enhancement for plugins' name checking [here](https://github.com/gmarik/Vundle.vim/compare/master...av3r4ge:plugin-names-security-enhancement).
What is your opinion?
Can it break current plugin names (can't see why names would contain other characters, but anyway)?
I think it can be a good start for preparing Vundle for this PR, security-wise.

**Comment by bcnjr5** [comment]
Merged it in. That commit fixes a lot of problems.

**Comment by bcnjr5** [comment]
`tig`'s output is getting interesting. I wonder just how many interconnected merges and branches it can draw.

**Comment by jdevera** [comment]
- We should not have an option to make vundle secure, all behaviour should be :)
- The fix for s:expand_path in `config.vim` I'd like to apply separately, I've asked @av3r4ge to create a separate PR.
- When we have dependency support it will always be disabled by default. I like the idea of having a global option and then being able to override it per plugin.
- I'm against asking anything to the user during installation, the user should speak their intentions with configuration.
- I don't want to encourage at all any tricks in the json files. For me, this should be as if vimscript had a proper json parser that was not evalling. If something doesn't look like a straight dep, then we don't consider it.

**Comment by bcnjr5** [comment]
@jdevera: Will do.

**Comment by bcnjr5** [comment]
I had stuff come up that has prevented me from working on this. Sorry.

**Pull request closed without merging**